### ISSUE TITLE
Session unscheduled disassembly when sending SIGINT to gdb session controlling a remote ARM target

### ISSUE THREAD

**Issue by mubes** [cs01/gdbgui#176]
##### Describe the issue
When sending SIGINT to the gdb controlling a remote target process the session crashes.

##### Steps to replicate

I created the issue using a BlackMagic Probe target driving an ARM Cortex-M3 using gcc.  On sending SIGINT from the browser the following response is generated. (Note; signals to the inferior have no meaning in this setup).

gdbgui -g ~/bin/armgcc/bin/arm-none-eabi-gdb ofiles/firmware.elf --project ~/Develop/dollshouse -n 
View gdbgui at http://127.0.0.1:5000
exit gdbgui by pressing CTRL+C
[2018-02-13 23:38:33,983] ERROR in app: Exception on /send_signal_to_pid [POST]
Traceback (most recent call last):
  File "/home/dmarples/.local/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/dmarples/.local/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/dmarples/.local/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/dmarples/.local/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/dmarples/.local/lib/python2.7/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/dmarples/.local/lib/python2.7/site-packages/gdbgui/backend.py", line 470, in send_signal_to_pid
    return jsonify({'message': 'sent signal %s (%s) to process id %s' % (signal_name, signal_obj.value, pid_str)})
AttributeError: 'int' object has no attribute 'value'

##### Describe your environment
* Operating system and version:
* gdbgui version (`gdbgui -v`): 0.11.1.2
* gdb version (`gdb -v`): GNU gdb (GNU Tools for ARM Embedded Processors 6-2017-q2-update) 7.12.1.20170417-git
* browser: Occurs with firefox and chrome
* python packages (`pip freeze`):
altgraph==0.14
asn1crypto==0.22.0
click==6.7
cryptography==1.9
Cython==0.25.2
dis3==0.1.1
duplicity===-version
enum34==1.1.6
fasteners==0.12.0
Flask==0.12.2
Flask-Compress==1.4.0
Flask-SocketIO==2.9.3
future==0.16.0
gdbgui==0.11.1.2
gevent==1.2.2
greenlet==0.4.13
gyp==0.1
idna==2.5
ipaddress==1.0.17
itsdangerous==0.24
Jinja2==2.10
keyring==10.4.0
keyrings.alt==2.2
macholib==1.8
MarkupSafe==1.0
mercurial==4.3.1
monotonic==1.0
mutagen==1.38
numpy==1.12.1
pefile==2017.11.5
pexpect==4.2.1
pycrypto==2.6.1
pygame===1.9.1release
pygdbmi==0.8.2.0
Pygments==2.2.0
pygobject==3.24.1
PyInstaller==3.3
pykaraoke==0.7.5
python-engineio==2.0.2
python-socketio==1.8.4
pyxdg==0.25
SecretStorage==2.3.1
six==1.11.0
vboxapi==1.0
Werkzeug==0.14.1
wxPython==3.0.2.0
wxPython-common==3.0.2.0

**Comment by cs01** [comment]
The title of this issue is confusing. Seems like it should simply be "Sending SIGING to gdb from gdbgui raises Python exception". Is that an accurate summary?

In any case, it looks like the `signal` module in the version of Python you're running does not contain what gdbgui expects. Could you re-run and print out `signal_obj` 
here? https://github.com/cs01/gdbgui/blob/master/gdbgui/backend.py#L470

Seems like its an integer rather than an object.

**Comment by mubes** [comment]
Yes, it was a not very good attempt at humor in the honor of the Falcon Heavy central booster.

Anyway, it is indeed an int;

```
2 <type 'int'>
```

my not very elegant work-around is;

```
    if (type(signal_obj)==int):
        return jsonify({'message': 'sent signal %s (%d) to process id %s' % (signal_name, signal_obj, pid_str)})
    else:
        return jsonify({'message': 'sent signal %s (%s) to process id %s' % (signal_name, signal_obj.value, pid_str)})
```

...which seems to work fine, resulting in;

```
Program received signal SIGINT, Interrupt.
0x08006932 in USB_LP_CAN1_RX0_IRQHandler () at thirdparty/USB/src/usb_istr.c:148
148              _SetISTR((uint16_t)CLR_ERR);

gdbgui noticed a signal was recieved (Interrupt, SIGINT).
If the program exited due to a fault, you can attempt to re-enter the state of the program when the fault 
occurred by clicking the below button.
```

Is there any chance of (a) hooking this SIGINT behavior to the run button (so it becomes a pause button and (b) Removing this warning for the case of SIGINT (Since it doesn't need to be re-connected).

You could also remove the send-to-inferior signal stuff in the UI for the case of an embedded target since it doesn't really mean much in that circumstance.

Great work BTW!!!

DAVE

**Comment by cs01** [comment]
> Would there be any chance of (a) hooking this SIGINT behavior to the run button (so it becomes a pause button and 

I don't want to modify the behavior of the run button. But Maybe it makes sense to add another button to easily send SIGINT whenever. Perhaps a "Pause" button?

> (b) Removing this warning for the case of SIGINT (Since it doesn't need to be re-connected)?

Yes I can do that.

> You could also remove the send-to-inferior signal stuff in the UI for the case of an embedded target since it doesn't really mean much in that circumstance.

Agreed, I will do this.

> Great work BTW!!!
Thank you!  💻

**Comment by mubes** [comment]
I think a pause button would be fine, but at any point in time only the run or pause button will be in operation so, assuming your code knows the current run state of the target, the one button could easily do both jobs with no increase in real-estate....however, the important thing is to get the functionality in there.

Regards

DAVE

**Comment by cs01** [comment]
```
return jsonify({'message': 'sent signal %s (%s) to process id %s' % (signal_name, signal_obj.value, pid_str)})
AttributeError: 'int' object has no attribute 'value'
```

Turns our python3 has the value attribute and python2 does not. This was brought up in another issue (#196 ) and fixed in bbc51d46.

Since you brought up a couple good features, I am going to close this issue and create separate issues for them.
